### PR TITLE
Native layer norm backwards

### PR DESCRIPTION
**Pull request by jjsjann123**
Fixes #{issue number}

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/kernel_cache.cpp [hunk 1]
```diff
@@ -566,7 +569,12 @@ std::vector<at::Tensor> FusionSegmentRuntime::runWithInput(
   // Produce final global output
   std::vector<IValue> fusion_outputs;
   for (auto output : segmented_fusion_->outputs()) {
-    fusion_outputs.push_back(tensor_map.at(output));
+    if (auto iter = tensor_map.find(output); iter != tensor_map.end()) {
+      fusion_outputs.push_back(iter->second);
+    } else {
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/kernel_cache.cpp:591]: @shmsong 
This is the temporary hack I put there. I think it's safe as long as we insert a check here that we are only creating empty tensor for dangling TV.
But does it make more sense to have fusion segmentation to keep track of that?

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 2]
```diff
@@ -989,22 +992,26 @@ class IrParser {
...
 
-            value_map.emplace(node->output(0)->unique(), grad_in);
-
-            // TODO: grad_bias and grad_weight are disabled because
-            // they are incompabilble with grad_in fusion
-            // Requires seperate kernels
+            if (output_mask[0]) {
+              value_map.emplace(node->output(0)->unique(), grad_in);
+            } else {
+              value_map.emplace(node->output(0)->unique(), nullptr);
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/parser.cpp]: note to myself. replace this with an empty tensor instead.

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 3]
```diff
@@ -989,22 +992,26 @@ class IrParser {
...
-            //  auto grad_weight = sum(mul(grad_out, x_hat),
-            //  outer_reduction_axes);
-            //  value_map.emplace(node->output(1)->unique(), grad_weight);
-            // }
+            if (output_mask[1] && weight != nullptr) {
+              auto grad_weight = sum(mul(grad_out, x_hat),
+              outer_reduction_axes);
+              value_map.emplace(node->output(1)->unique(), grad_weight);
+            } else {
+              value_map.emplace(node->output(1)->unique(), TensorViewBuilder().build());
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/parser.cpp]: @kevinstephano @shmsong 
So this is the place where empty tensor was created.

File: test/test_jit_cuda_fuser.py [hunk 4]
```diff
@@ -995,6 +995,80 @@ def test_reduction(self):
...
+            return o
+        def t(shapes: List[int], x, eps: float, cudnn: bool):
+            o = torch.layer_norm(x, shapes, None, None, eps, cudnn)
+            o = torch.relu(o)
+            return o
+        model = {3 : t_wb, 2 : t_w, 1 : t_b, 0: t}
+
+        for w, b in itertools.product([True, False], repeat=2):
+            if w == False and b == True:
+                # TODO: temporarily WAR failing codegen;
```
Comment (jjsjann123) [inline test/test_jit_cuda_fuser.py]: Note to myself: create an issue. This is where codegen fails.

File: torch/csrc/jit/codegen/cuda/shape_inference.cpp [hunk 5]
```diff
@@ -318,6 +318,27 @@ class NaiveTypePropagator {
         break;
       }
       default:
+
+        // TODO: temporary placement to avoid recompilation
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/shape_inference.cpp]: Note to myself: fix this

File: torch/csrc/jit/codegen/cuda/kernel_cache.cpp (hunk 1 above)
Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/kernel_cache.cpp:591]: I see. I can add the dangling output TV's back into the segmented Fusion. The way it is now should work either way. Please feel free to add any additional checks that you'd like to do here and I can migrate that later.

Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/kernel_cache.cpp:591]: Thanks, I'll update this when merging fixes from dev_branch here.

File: test/test_jit_cuda_fuser.py [hunk 6]
```diff
@@ -995,6 +995,80 @@ def test_reduction(self):
                         perm1 = range(len(x))
                         self._reduction_helper(x, axes, torch.float32, "cuda", perm0, perm1, keepdim)
 
+    def _layer_norm_autodiff_helper(self, model, grad, shapes, args):
+        jit_model = torch.jit.script(model)
+
+        eps = np.random.random() * 1e-4
+        use_cudnn = bool(np.random.randint(0, 2))
```
Comment (csarofeen) [inline test/test_jit_cuda_fuser.py:1002]: Should we just run both?

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 7]
```diff
@@ -989,22 +992,29 @@ class IrParser {
                 unaryOp(UnaryOpType::Reciprocal, num_features);
             auto* grad_in = mul(mul(reciprocal_size, rstd), inner);
 
-            value_map.emplace(node->output(0)->unique(), grad_in);
-
-            // TODO: grad_bias and grad_weight are disabled because
-            // they are incompabilble with grad_in fusion
-            // Requires seperate kernels
+            if (output_mask[0]) {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/parser.cpp:995]: Seems like you're inserting dummy 0-dim tensors in place when there aren't certain outputs required. Is that true? Is this the dangling output you're referring to for segmentation?

**Review by csarofeen — APPROVED** [review]
LGTM

File: test/test_jit_cuda_fuser.py (hunk 6 above)
Comment (jjsjann123) [reply test/test_jit_cuda_fuser.py:1002]: I wouldn't worry too much about it yet. We are currently ignoring the cudnn flag, so does pytorch :)
https://github.com/pytorch/pytorch/blob/3721962c33a64946bd3be8ce8c032c4e7d895d22/aten/src/ATen/native/layer_norm.cpp#L99

File: torch/csrc/jit/codegen/cuda/parser.cpp (hunk 7 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/parser.cpp:995]: Yes. This is the portion that's causing earlier segfault during the execution of segmented fusion. Now it should be working fine.

**Comment by jjsjann123** [comment]
local tests passed. Will merge when CI turns green.

**Pull request merged**